pkg/controller/reconciler/deploymenttemplate_reconciler_test.go

```diff
@@ -0,0 +1,401 @@
+		assert.Equal(t, status.ObservedGeneration, current.Generation, "Status is not updated")
+
+		if assert.Equal(t, radappiov1alpha3.DeploymentTemplatePhraseUpdating, current.Status.Phrase) {
+			assert.NotEmpty(t, current.Status.Operation)
+			assert.NotEqual(t, oldOperation, current.Status.Operation)
+		}
+
+	}, deploymentTemplateTestWaitDuration, deploymentTemplateTestWaitInterval, "failed to enter updating state")
+
+	return status
+}
+
+func waitForDeploymentTemplateStateReady(t *testing.T, client client.Client, name types.NamespacedName) *radappiov1alpha3.DeploymentTemplateStatus {
+	ctx := testcontext.New(t)
+
+	logger := t
+	status := &radappiov1alpha3.DeploymentTemplateStatus{}
+	require.EventuallyWithTf(t, func(t *assert.CollectT) {
+		logger.Logf("Fetching DeploymentTemplate: %+v", name)
+		current := &radappiov1alpha3.DeploymentTemplate{}
+		err := client.Get(ctx, name, current)
+		require.NoError(t, err)
+
+		status = &current.Status
+		logger.Logf("DeploymentTemplate.Status: %+v", current.Status)
+		assert.Equal(t, status.ObservedGeneration, current.Generation, "Status is not updated")
+
+		if assert.Equal(t, radappiov1alpha3.DeploymentTemplatePhraseReady, current.Status.Phrase) {
+			assert.Empty(t, current.Status.Operation)
+		}
+	}, deploymentTemplateTestWaitDuration, deploymentTemplateTestWaitInterval, "failed to enter ready state")
+
+	return status
+}
+
+func waitForDeploymentTemplateStateDeleting(t *testing.T, client client.Client, name types.NamespacedName, oldOperation *radappiov1alpha3.ResourceOperation) *radappiov1alpha3.DeploymentTemplateStatus {
+	ctx := testcontext.New(t)
+
+	logger := t
+	status := &radappiov1alpha3.DeploymentTemplateStatus{}
+	require.EventuallyWithTf(t, func(t *assert.CollectT) {
+		logger.Logf("Fetching DeploymentTemplate: %+v", name)
+		current := &radappiov1alpha3.DeploymentTemplate{}
+		err := client.Get(ctx, name, current)
+		assert.NoError(t, err)
+
+		status = &current.Status
+		logger.Logf("DeploymentTemplate.Status: %+v", current.Status)
+		assert.Equal(t, status.ObservedGeneration, current.Generation, "Status is not updated")
+
+		assert.Equal(t, radappiov1alpha3.DeploymentTemplatePhraseDeleting, current.Status.Phrase)
+	}, deploymentTemplateTestWaitDuration, deploymentTemplateTestWaitInterval, "failed to enter deleting state")
+
+	return status
+}
+
+func waitForDeploymentTemplateStateDeleted(t *testing.T, client client.Client, name types.NamespacedName) {
+	ctx := testcontext.New(t)
+
+	logger := t
+	require.Eventuallyf(t, func() bool {
+		logger.Logf("Fetching DeploymentTemplate: %+v", name)
+		current := &radappiov1alpha3.DeploymentTemplate{}
+		err := client.Get(ctx, name, current)
+		if apierrors.IsNotFound(err) {
+			return true
+		}
+
+		logger.Logf("DeploymentTemplate.Status: %+v", current.Status)
+		return false
+
+	}, deploymentTemplateTestWaitDuration, deploymentTemplateTestWaitInterval, "DeploymentTemplate still exists")
+}
```

Comment:
What's your plan for testing the more complex deletion scenarios?

Reply:
I added a couple of new functional tests, one that deploys a recipe with kubernetes resources via recipe. do you think we should expand these tests? any ideas for what to add?